File: camera/src/main/java/com/dastanapps/camera/Camera1.java
package com.dastanapps.camera;

import android.app.Activity;
import android.content.Context;
import android.graphics.SurfaceTexture;
import android.hardware.Camera;
import android.media.CamcorderProfile;
import android.media.MediaRecorder;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.v4.util.Pair;
import android.util.Log;
import android.view.MotionEvent;
import android.view.OrientationEventListener;
import android.view.Surface;
import android.widget.Toast;

import com.dastanapps.camera.listeners.Cam1OrientationEventListener;
import com.dastanapps.camera.listeners.Cam1SurfaceTextureListener;
import com.dastanapps.camera.view.Cam1AutoFitTextureView;
import com.dastanapps.camera.view.FocusImageView;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

/**
 * Created by dastaniqbal on 18/01/2018.
 * dastanIqbal@marvelmedia.com
 * 18/01/2018 6:32
 */

public class Camera1 {
    private static final String TAG = "Camera1";
    public final static int FOCUSING_FOCUS_VIEW = 0;
    public final static int SUCCESS_FOCUS_VIEW = 1;
    public final static int FAILED_FOCUS_VIEW = 2;
    private final ICamera1 mCamera1Listener;
    private Cam1SurfaceTextureListener mCameraSurfaceTextureListener;
    private OrientationEventListener mOrientationEventListener;
    private Cam1AutoFitTextureView mTextureView;
    private Context mContext;
    private Activity mActivity;

    private Camera mCamera;
    private int mDisplayOrientation;
    private int cameraId;
    private Camera.Parameters mCharacteristics;
    private Camera.Size mVideoSize;
    private Camera.Size mPreviewSize;
    private String mNextVideoAbsolutePath;

    /**
     * MediaRecorder
     */
    private MediaRecorder mMediaRecorder;

    /**
     * Whether the app is recording video now
     */
    private boolean mIsRecordingVideo;

    /**
     * A {@link Semaphore} to prevent the app from exiting before closing the camera.
     */
    private Semaphore mCameraOpenCloseLock = new Semaphore(1);
    private FocusImageView focusImage;
    private Handler mainHandler = new Handler(Looper.getMainLooper()) {

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            switch (msg.what) {
                case FOCUSING_FOCUS_VIEW:
                    MotionEvent touchEvent = (MotionEvent) msg.obj;
                    if (touchEvent != null && focusImage != null) {
                        focusImage.startFocusing(touchEvent);
                    }
                    break;
                case SUCCESS_FOCUS_VIEW:
                    if (focusImage != null) {
                        focusImage.focusSuccess();
                    }
                    break;
                case FAILED_FOCUS_VIEW:
                    if (focusImage != null) {
                        focusImage.focusFailed();
                    }
                    break;
            }
        }
    };


    public Camera1(Context context, Cam1AutoFitTextureView mTextureView, ICamera1 camera1Listener) {
        this.mContext = context;
        this.mActivity = (Activity) context;
        this.mTextureView = mTextureView;
        this.mCamera1Listener = camera1Listener;
        mCameraSurfaceTextureListener = new Cam1SurfaceTextureListener(this, mTextureView, mActivity);
        this.mTextureView.setSurfaceTextureListener(mCameraSurfaceTextureListener);
        this.mTextureView.setMainHandler(mainHandler);

        mOrientationEventListener = new Cam1OrientationEventListener(mContext);
        setupManager();
    }

    private void setupManager() {
        Pair<Camera.CameraInfo, Integer> backCamera = Camera1Helper.getCameraInfo(Camera.CameraInfo.CAMERA_FACING_BACK);
        cameraId = backCamera.second;
        mDisplayOrientation = backCamera.first.orientation;
    }

    protected void onResume() {
        if (mOrientationEventListener.canDetectOrientation()) {
            mOrientationEventListener.enable();
        }
    }

    protected void onPause() {
        releaseMediaRecorder();
        closeCamera();
        mOrientationEventListener.disable();
    }

    public void openCamera() {
        closeCamera();

        try {
            if (!mCameraOpenCloseLock.tryAcquire(2500, TimeUnit.MILLISECONDS)) {
                throw new RuntimeException("Time out waiting to lock camera opening.");
            }
            mCamera = Camera.open(cameraId);
            mTextureView.setCameraSettings(mCamera);
            mCharacteristics = mCamera.getParameters();
            mVideoSize = Camera1Helper.chooseVideoSize(mCharacteristics.getSupportedVideoSizes());
            mPreviewSize = Camera1Helper.chooseOptimalSize(mCharacteristics.getSupportedPreviewSizes(),
                    mTextureView.getWidth(), mTextureView.getHeight(), mVideoSize);
            int orientation = mContext.getResources().getConfiguration().orientation;
//        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
//            mTextureView.setAspectRatio(mPreviewSize.width, mPreviewSize.height);
//        } else {
//            mTextureView.setAspectRatio(mPreviewSize.height, mPreviewSize.width);
//        }
//            Camera1Helper.configureTransform(mActivity, mPreviewSize, mTextureView, mTextureView.getWidth(), mTextureView.getHeight());
//            mCharacteristics.setPreviewSize(mPreviewSize.width, mPreviewSize.height);
//            mCamera.setParameters(mCharacteristics);

            try {
                SurfaceTexture texture = mTextureView.getSurfaceTexture();
                if (texture != null) {
                    texture.setDefaultBufferSize(mPreviewSize.width, mPreviewSize.height);
                }
                mCamera.setDisplayOrientation(Camera1Helper.setDisplayOrientation(mActivity, mDisplayOrientation));
                mCamera.setPreviewTexture(mTextureView.getSurfaceTexture());
                mCamera.startPreview();
            } catch (IOException ioe) {
                // Something bad happened
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void setAutoExposure(int autoExposure) {
        mCharacteristics.setAutoExposureLock(false);
        int maxmax = mCharacteristics.getMaxExposureCompensation();
        int minmin = mCharacteristics.getMinExposureCompensation();
        int all = (-minmin) + maxmax;
        int time = 100 / all;
        int ae = ((autoExposure / time) - maxmax) > maxmax ? maxmax : ((autoExposure / time) - maxmax) < minmin ? minmin : ((autoExposure / time) - maxmax);
        mCharacteristics.setExposureCompensation(ae);
        mCamera.setParameters(mCharacteristics);
        mCharacteristics.setAutoExposureLock(true);
    }

    public void closeCamera() {
        try {
            mCameraOpenCloseLock.release();
            mCameraOpenCloseLock.acquire();
            if (mCamera != null) {
                mCamera.stopPreview();
                mCamera.release();
                mCamera = null;
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        } finally {
            mCameraOpenCloseLock.release();
        }
    }

    public boolean isCameraOpen() {
        return mCamera != null;
    }

    public void switchFaces() {
        if (cameraId == Camera.CameraInfo.CAMERA_FACING_BACK)
            cameraId = Camera.CameraInfo.CAMERA_FACING_FRONT; //Front
        else cameraId = Camera.CameraInfo.CAMERA_FACING_BACK; //Back
        openCamera();
    }

    public DialogHelper.FeatureListDialog toggleFlash() {
        final List<String> flashes = mCharacteristics.getSupportedFlashModes();
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(flashes, new DialogHelper.FeatureListDialog.ISelectItem() {
            @Override
            public void onSelectItem(int which) {
                mCharacteristics.setFlashMode(flashes.get(which));
                mCamera.setParameters(mCharacteristics);
            }
        });
    }

    public DialogHelper.FeatureListDialog showScenesDialog() {
        final List<String> sceneModes = mCharacteristics.getSupportedSceneModes();
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(sceneModes, new DialogHelper.FeatureListDialog.ISelectItem() {
            @Override
            public void onSelectItem(int which) {
                mCharacteristics.setSceneMode(sceneModes.get(which));
                mCamera.setParameters(mCharacteristics);
            }
        });
    }

    public DialogHelper.FeatureListDialog showEffectsDialog() {
        final List<String> colorEffects = mCharacteristics.getSupportedColorEffects();
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(colorEffects, new DialogHelper.FeatureListDialog.ISelectItem() {
            @Override
            public void onSelectItem(int which) {
                mCharacteristics.setColorEffect(colorEffects.get(which));
                mCamera.setParameters(mCharacteristics);
            }
        });
    }

    public DialogHelper.FeatureListDialog showWhiteBalance() {
        final List<String> colorEffects = mCharacteristics.getSupportedWhiteBalance();
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(colorEffects, new DialogHelper.FeatureListDialog.ISelectItem() {
            @Override
            public void onSelectItem(int which) {
                mCharacteristics.setWhiteBalance(colorEffects.get(which));
                mCamera.setParameters(mCharacteristics);
            }
        });
    }

    public void setFocusImage(@NonNull FocusImageView focusImage) {
        this.focusImage = focusImage;
    }

    public void toggleRecording() {
        if (mIsRecordingVideo) {
            stopRecordingVideo();
            releaseMediaRecorder();
            mCamera.lock();
        } else if (prepareMediaRecorder()) {
            startRecordingVideo();
        } else {
            releaseMediaRecorder();
        }
    }

    public void startRecordingVideo() {
        if (!mTextureView.isAvailable() || null == mPreviewSize) {
            return;
        }
        mMediaRecorder.start();
        mIsRecordingVideo = true;
    }

    private boolean prepareMediaRecorder() {
        if (null == mActivity) {
            return false;
        }
        Surface surface = new Surface(mTextureView.getSurfaceTexture());
        if (mMediaRecorder == null)
            mMediaRecorder = new MediaRecorder();

        // Step 1: Unlock and set camera to MediaRecorder
        mCamera.unlock();
        mMediaRecorder.setCamera(mCamera);

        // Step 2: Set sources
        mMediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
        mMediaRecorder.setVideoSource(MediaRecorder.VideoSource.CAMERA);


        // Step 3: Set a CamcorderProfile (requires API Level 8 or higher)
        // Customise your profile based on a pre-existing profile
        CamcorderProfile profile = getBaseRecordingProfile();
        profile.fileFormat = MediaRecorder.OutputFormat.MPEG_4;
        profile.videoCodec = MediaRecorder.VideoEncoder.H264;
        profile.audioCodec = MediaRecorder.AudioEncoder.AAC;
        profile.videoFrameHeight = mVideoSize.height;
        profile.videoFrameWidth = mVideoSize.width;
        profile.videoFrameRate = 30;
        profile.videoBitRate = 5000000;

        mMediaRecorder.setOrientationHint(Camera1Helper.setDisplayOrientation(mActivity, mDisplayOrientation));
        ;
        mMediaRecorder.setProfile(profile);
//
        if (mNextVideoAbsolutePath == null || mNextVideoAbsolutePath.isEmpty()) {
            mNextVideoAbsolutePath = getVideoFilePath(mActivity);
        }
        // Step 4: Set output file
        mMediaRecorder.setOutputFile(mNextVideoAbsolutePath);

        // Step 5: Set the preview output
        mMediaRecorder.setPreviewDisplay(surface);

        // Step 6: Prepare configured MediaRecorder
        try {
            mMediaRecorder.prepare();
        } catch (Exception e) {
            Log.d(TAG, "IllegalStateException preparing MediaRecorder: " + e.getMessage());
            releaseMediaRecorder();
            Toast.makeText(mActivity, "Video Prepare Failed", Toast.LENGTH_LONG).show();
            return false;
        }
        return true;
    }

    private CamcorderProfile getBaseRecordingProfile() {
        CamcorderProfile returnProfile;
        if (CamcorderProfile.hasProfile(CamcorderProfile.QUALITY_720P)) {
            returnProfile = CamcorderProfile.get(CamcorderProfile.QUALITY_720P);
        } else if (CamcorderProfile.hasProfile(CamcorderProfile.QUALITY_480P)) {
            returnProfile = CamcorderProfile.get(CamcorderProfile.QUALITY_480P);
        } else {
            returnProfile = getDefaultRecordingProfile();
        }
        return returnProfile;
    }

    private CamcorderProfile getDefaultRecordingProfile() {
        CamcorderProfile highProfile = CamcorderProfile.get(CamcorderProfile.QUALITY_HIGH);
        if (highProfile != null) {
            return highProfile;
        }
        CamcorderProfile lowProfile = CamcorderProfile.get(CamcorderProfile.QUALITY_LOW);
        if (lowProfile != null) {
            return lowProfile;
        }
        throw new RuntimeException("No quality level found");
    }

    private String getVideoFilePath(Context context) {
        final File dir = context.getExternalFilesDir(null);
        return (dir == null ? "" : (dir.getAbsolutePath() + "/"))
                + System.currentTimeMillis() + ".mp4";
    }

    private Handler stopRecodeingHandler = new Handler();
    private Runnable stopRecodingRunnable = new Runnable() {
        @Override
        public void run() {
            if (mMediaRecorder != null)
                mMediaRecorder.stop();
        }
    };

    private void releaseMediaRecorder() {
        if (mMediaRecorder != null) {
            mMediaRecorder.reset();   // clear recorder configuration
            mMediaRecorder.release(); // release the recorder object
            mMediaRecorder = null;
            mCamera.lock();           // lock camera for later use
        }
    }

    private void stopRecordingVideo() {
        // UI
        mIsRecordingVideo = false;

        stopRecodeingHandler.removeCallbacks(stopRecodingRunnable);
        stopRecodeingHandler.postDelayed(stopRecodingRunnable, 100);

        if (null != mActivity) {
            Toast.makeText(mActivity, "Video saved: " + mNextVideoAbsolutePath,
                    Toast.LENGTH_SHORT).show();
            Log.d(TAG, "Video saved: " + mNextVideoAbsolutePath);
        }
        mNextVideoAbsolutePath = null;
        mCamera1Listener.cameraRecordingStopped();
    }
}

File: camera/src/main/java/com/dastanapps/camera/Camera1.kt
package com.dastanapps.camera

import android.app.Activity
import android.content.Context
import android.graphics.SurfaceTexture
import android.hardware.Camera
import android.media.CamcorderProfile
import android.media.MediaRecorder
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import android.view.MotionEvent
import android.view.OrientationEventListener
import android.view.Surface
import android.widget.Toast
import com.dastanapps.CameraHelper
import com.dastanapps.camera.listeners.Cam1OrientationEventListener
import com.dastanapps.camera.listeners.Cam1SurfaceTextureListener
import com.dastanapps.camera.view.Cam1AutoFitTextureView
import com.dastanapps.camera.view.FocusImageView
import com.dastanapps.camera2.view.CameraSurfaceRenderer
import com.dastanapps.encoder.MediaVideoEncoder
import java.io.IOException
import java.util.concurrent.Semaphore
import java.util.concurrent.TimeUnit

/**
 * Created by dastaniqbal on 18/01/2018.

 * 18/01/2018 6:32
 */
@SuppressWarnings("deprecation")
class Camera1(private val mContext: Context, private val mTextureView: Cam1AutoFitTextureView, private val mCamera1Listener: ICamera1) {
    private val mCameraSurfaceTextureListener: Cam1SurfaceTextureListener
    private val mOrientationEventListener: OrientationEventListener
    private val mActivity: Activity?

    private var mCamera: Camera? = null
    private var mDisplayOrientation: Int = 0
    private var cameraId: Int = 0
    private var mCharacteristics: Camera.Parameters? = null
    private var mVideoSize: Camera.Size? = null
    private var mPreviewSize: Camera.Size? = null
    private var mNextVideoAbsolutePath: String? = null
    private var filterTexture: SurfaceTexture? = null
    /**
     * MediaRecorder
     */
    private var mMediaRecorder: MediaRecorder? = null

    /**
     * Whether the app is recording video now
     */
    private var mIsRecordingVideo: Boolean = false

    /**
     * A [Semaphore] to prevent the app from exiting before closing the camera.
     */
    private val mCameraOpenCloseLock = Semaphore(1)
    private var focusImage: FocusImageView? = null
    private var screenCurrentRotation: Int = 0
    private val mainHandler = object : Handler(Looper.getMainLooper()) {

        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            when (msg.what) {
                FOCUSING_FOCUS_VIEW -> {
                    val touchEvent = msg.obj as MotionEvent
                    if (touchEvent != null && focusImage != null) {
                        focusImage!!.startFocusing(touchEvent)
                    }
                }
                SUCCESS_FOCUS_VIEW -> if (focusImage != null) {
                    focusImage!!.focusSuccess()
                }
                FAILED_FOCUS_VIEW -> if (focusImage != null) {
                    focusImage!!.focusFailed()
                }
                3 -> {
                    screenCurrentRotation = msg.obj as Int
                    mCamera1Listener.orientationChanged(screenCurrentRotation)
                }
            }
        }
    }

    val isCameraOpen: Boolean
        get() = mCamera != null

    val rotationCorrection: Int
        get() {
            val displayRotation = mDisplayOrientation * 90
            val mCamInfo = Camera.CameraInfo()
            Camera.getCameraInfo(cameraId, mCamInfo)

            if (cameraId == Camera.CameraInfo.CAMERA_FACING_FRONT) {
                val mirroredRotation = (mCamInfo.orientation + displayRotation) % 360
                return (360 - mirroredRotation) % 360
            } else {
                return (mCamInfo.orientation - displayRotation + 360) % 360
            }
        }

    private val stopRecodeingHandler = Handler()
    private val stopRecodingRunnable = Runnable {
        if (mMediaRecorder != null)
            mMediaRecorder!!.stop()
    }


    init {
        this.mActivity = mContext as Activity
        mCameraSurfaceTextureListener = Cam1SurfaceTextureListener(this, mTextureView, mActivity)
        // this.mTextureView.surfaceTextureListener = mCameraSurfaceTextureListener
        this.mTextureView.setMainHandler(mainHandler)

        mOrientationEventListener = Cam1OrientationEventListener(mContext, mainHandler)
        setupManager()
        setUpFilters()
    }

    private fun setupManager() {
        val backCamera = Camera1Helper.getCameraInfo(Camera.CameraInfo.CAMERA_FACING_BACK)
        cameraId = backCamera!!.second
        mDisplayOrientation = backCamera.first.orientation
    }

    fun onResume() {
        if (mOrientationEventListener.canDetectOrientation()) {
            mOrientationEventListener.enable()
        }
    }

    fun onPause() {
        releaseMediaRecorder()
        closeCamera()
        mOrientationEventListener.disable()
    }

    private lateinit var camCoderProfile: CamcorderProfile

    fun openCamera() {
        closeCamera()

        try {
            if (!mCameraOpenCloseLock.tryAcquire(2500, TimeUnit.MILLISECONDS)) {
                throw RuntimeException("Time out waiting to lock camera opening.")
            }
            mCamera = Camera.open(cameraId)
            mCharacteristics = mCamera!!.parameters
            mTextureView.setCameraSettings(mCamera!!)

            camCoderProfile = Camera1Helper.baseRecordingProfile(cameraId)

            val videoSizeList = when {
                mCharacteristics!!.supportedVideoSizes != null -> mCharacteristics!!.supportedVideoSizes
                else -> mCharacteristics!!.supportedPreviewSizes
            }

            mVideoSize = Camera1Helper.getSizeWithClosestRatio(videoSizeList, camCoderProfile.videoFrameWidth, camCoderProfile.videoFrameHeight)
            mPreviewSize = Camera1Helper.getSizeWithClosestRatio(mCharacteristics!!.supportedPreviewSizes, mVideoSize!!.width, mVideoSize!!.height)
            //Camera1Helper.chooseOptimalSize(mCharacteristics!!.supportedPreviewSizes, mTextureView.width, mTextureView.height, mVideoSize!!)

            val orientation = mContext.resources.configuration.orientation
            //        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            //            mTextureView.setAspectRatio(mPreviewSize.width, mPreviewSize.height);
            //        } else {
            //            mTextureView.setAspectRatio(mPreviewSize.height, mPreviewSize.width);
            //        }
            //            Camera1Helper.configureTransform(mActivity, mPreviewSize, mTextureView, mTextureView.getWidth(), mTextureView.getHeight());
//            mCharacteristics!!.setPreviewSize(1280, 960)
//            mCamera!!.parameters = mCharacteristics;

            try {
//                val texture = mTextureView.surfaceTexture
//                if (mPreviewSize != null)
//                    texture?.setDefaultBufferSize(mPreviewSize!!.width, mPreviewSize!!.height)
                mCamera!!.setDisplayOrientation(Camera1Helper.setDisplayOrientation(cameraId, mActivity, mDisplayOrientation))
                mCharacteristics!!.focusMode = Camera.Parameters.FOCUS_MODE_AUTO
                mCharacteristics!!.flashMode = Camera.Parameters.FLASH_MODE_AUTO
                mCamera!!.parameters = mCharacteristics!!
                mCamera!!.setPreviewTexture(filterTexture);//mTextureView.surfaceTexture)
                mCamera!!.startPreview()
            } catch (ioe: IOException) {
                // Something bad happened
            }

        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }

    fun startPreview() {

    }

    fun setAutoExposure(autoExposure: Int) {
        mCharacteristics!!.autoExposureLock = false
        val maxmax = mCharacteristics!!.maxExposureCompensation
        val minmin = mCharacteristics!!.minExposureCompensation
        val all = -minmin + maxmax
        val time = 100 / all
        val ae = if (autoExposure / time - maxmax > maxmax) maxmax else if (autoExposure / time - maxmax < minmin) minmin else autoExposure / time - maxmax
        mCharacteristics!!.exposureCompensation = ae
        mCamera!!.parameters = mCharacteristics
        mCharacteristics!!.autoExposureLock = true
    }

    fun closeCamera() {
        try {
            mCameraOpenCloseLock.release()
            mCameraOpenCloseLock.acquire()
            if (mCamera != null) {
                mCamera!!.stopPreview()
                mCamera!!.release()
                mCamera = null
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } finally {
            mCameraOpenCloseLock.release()
        }
    }

    fun switchFaces() {
        if (cameraId == Camera.CameraInfo.CAMERA_FACING_BACK)
            cameraId = Camera.CameraInfo.CAMERA_FACING_FRONT //Front
        else
            cameraId = Camera.CameraInfo.CAMERA_FACING_BACK //Back
        openCamera()
    }

    fun toggleFlash(): DialogHelper.FeatureListDialog {
        val flashes = mCharacteristics!!.supportedFlashModes
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(flashes, object : DialogHelper.FeatureListDialog.ISelectItem {
            override fun onSelectItem(which: Int) {
                mCharacteristics!!.flashMode = flashes[which]
                mCamera!!.parameters = mCharacteristics
            }
        })
    }

    fun showScenesDialog(): DialogHelper.FeatureListDialog {
        val sceneModes = mCharacteristics!!.supportedSceneModes
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(sceneModes, object : DialogHelper.FeatureListDialog.ISelectItem {
            override fun onSelectItem(which: Int) {
                mCharacteristics!!.sceneMode = sceneModes[which]
                mCamera!!.parameters = mCharacteristics
            }
        })
    }

    fun showEffectsDialog(): DialogHelper.FeatureListDialog {
        val colorEffects = mCharacteristics!!.supportedColorEffects
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(colorEffects, object : DialogHelper.FeatureListDialog.ISelectItem {
            override fun onSelectItem(which: Int) {
                mCharacteristics!!.colorEffect = colorEffects[which]
                mCamera!!.parameters = mCharacteristics
            }
        })
    }

    fun showWhiteBalance(): DialogHelper.FeatureListDialog {
        val colorEffects = mCharacteristics!!.supportedWhiteBalance
        return DialogHelper.FeatureListDialog.newInstance().setFeatures(colorEffects, object : DialogHelper.FeatureListDialog.ISelectItem {
            override fun onSelectItem(which: Int) {
                mCharacteristics!!.whiteBalance = colorEffects[which]
                mCamera!!.parameters = mCharacteristics
            }
        })
    }

    fun setFocusImage(focusImage: FocusImageView) {
        this.focusImage = focusImage
    }

    fun toggleRecording() {
        when {
            mIsRecordingVideo -> {
                stopRecordingVideo()
                releaseMediaRecorder()
                mCamera!!.lock()
            }
            prepareMediaRecorder() -> startRecordingVideo()
            else -> releaseMediaRecorder()
        }
    }

    fun startRecordingVideo() {
        if (!mTextureView.isAvailable) {
            return
        }
        mMediaRecorder!!.start()
        mIsRecordingVideo = true
    }

    private fun prepareMediaRecorder(): Boolean {
        if (null == mActivity) {
            return false
        }
        val surface = Surface(mTextureView.surfaceTexture)
        mMediaRecorder = MediaRecorder()
        mMediaRecorder!!.setOnInfoListener { mr, what, extra ->
            when (what) {
                MediaRecorder.MEDIA_RECORDER_INFO_UNKNOWN -> {
                }
                MediaRecorder.MEDIA_RECORDER_INFO_MAX_DURATION_REACHED -> stopRecordingVideo()
                MediaRecorder.MEDIA_RECORDER_INFO_MAX_FILESIZE_REACHED -> stopRecordingVideo()
                else -> {
                }
            }// NOP
        }

        // Step 1: Unlock and set camera to MediaRecorder
        mCamera!!.unlock()
        mMediaRecorder!!.setCamera(mCamera)

        // Step 2: Set sources
        mMediaRecorder!!.setAudioSource(MediaRecorder.AudioSource.CAMCORDER)
        mMediaRecorder!!.setVideoSource(MediaRecorder.VideoSource.CAMERA)


        // Step 3: Set a CamcorderProfile (requires API Level 8 or higher)
        // Customise your profile based on a pre-existing profile

//        val profile = Camera1Helper.baseRecordingProfile(cameraId)
//        profile.fileFormat = MediaRecorder.OutputFormat.MPEG_4
//        profile.videoCodec = MediaRecorder.VideoEncoder.H264
//        profile.audioCodec = MediaRecorder.AudioEncoder.AAC
//
//        if (mVideoSize != null) {
//            profile.videoFrameHeight = 960 // mVideoSize!!.height
//            profile.videoFrameWidth = 1280 // mVideoSize!!.width
//        }
//        profile.videoFrameRate = 30
//        profile.videoBitRate = 9000000
//        profile.duration = 30
//
//        if (mVideoSize != null) {
//            mMediaRecorder!!.setProfile(profile);//CamcorderProfile.get(CamcorderProfile.QUALITY_720P));//profile)
//        } else {
//            mMediaRecorder!!.setProfile(CamcorderProfile.get(CamcorderProfile.QUALITY_HIGH))
//        }

        mMediaRecorder!!.setOutputFormat(camCoderProfile.fileFormat)
        mMediaRecorder!!.setVideoFrameRate(camCoderProfile.videoFrameRate)
        mMediaRecorder!!.setVideoSize(mVideoSize!!.width, mVideoSize!!.height)
        mMediaRecorder!!.setVideoEncodingBitRate(camCoderProfile.videoBitRate)
        mMediaRecorder!!.setVideoEncoder(camCoderProfile.videoCodec)

        mMediaRecorder!!.setAudioEncodingBitRate(camCoderProfile.audioBitRate)
        mMediaRecorder!!.setAudioChannels(camCoderProfile.audioChannels)
        mMediaRecorder!!.setAudioSamplingRate(camCoderProfile.audioSampleRate)
        mMediaRecorder!!.setAudioEncoder(camCoderProfile.audioCodec)

        if (mNextVideoAbsolutePath == null || mNextVideoAbsolutePath!!.isEmpty()) {
            mNextVideoAbsolutePath = getVideoFilePath(mActivity)
        }
        // Step 4: Set output file
        mMediaRecorder!!.setOutputFile(mNextVideoAbsolutePath)
        mMediaRecorder!!.setOrientationHint(CameraHelper.DEFAULT_ORIENTATIONS.get(screenCurrentRotation))

        // Step 5: Set the preview output
        //mMediaRecorder!!.setPreviewDisplay(surface) //Not required for TextureView

        // Step 6: Prepare configured MediaRecorder
        try {
            mMediaRecorder!!.prepare()
        } catch (e: Exception) {
            Log.d(TAG, "IllegalStateException preparing MediaRecorder: " + e.message)
            releaseMediaRecorder()
            Toast.makeText(mActivity, "Video Prepare Failed", Toast.LENGTH_LONG).show()
            return false
        }

        return true
    }

    private fun getVideoFilePath(context: Context): String {
        val dir = context.getExternalFilesDir(null)
        return ((if (dir == null) "" else dir.absolutePath + "/")
                + System.currentTimeMillis() + ".mp4")
    }

    private fun releaseMediaRecorder() {
        if (mMediaRecorder != null) {
            mMediaRecorder!!.reset()   // clear recorder configuration
            mMediaRecorder!!.release() // release the recorder object
            mMediaRecorder = null
            mCamera!!.lock()           // lock camera for later use
        }
    }

    private fun stopRecordingVideo() {
        // UI
        mIsRecordingVideo = false

        stopRecodeingHandler.removeCallbacks(stopRecodingRunnable)
        stopRecodeingHandler.postDelayed(stopRecodingRunnable, 100)

        if (null != mActivity) {
            Toast.makeText(mActivity, "Video saved: " + mNextVideoAbsolutePath!!,
                    Toast.LENGTH_SHORT).show()
            Log.d(TAG, "Video saved: " + mNextVideoAbsolutePath!!)
        }
        mNextVideoAbsolutePath = null
        mCamera1Listener.cameraRecordingStopped()
    }

    var cameraSurfaceRenderer: CameraSurfaceRenderer? = null

    fun setVideoEncoder(encoder: MediaVideoEncoder) {
        cameraSurfaceRenderer!!.setVideoEnocder(mTextureView, encoder)
    }

    private fun setUpFilters() {
        cameraSurfaceRenderer = CameraSurfaceRenderer()
        cameraSurfaceRenderer?.setListener({ surfaceTexture ->
            filterTexture = surfaceTexture
            Handler(Looper.getMainLooper()).post {
                openCamera()
            }
        })
        mTextureView.setRenderer(cameraSurfaceRenderer)
    }

    fun changeFilter() {
        if (cameraSurfaceRenderer != null) {
            mTextureView.queueEvent {
                cameraSurfaceRenderer?.changeFilter()
            }
        }
    }

    companion object {
        private val TAG = "Camera1"
        val FOCUSING_FOCUS_VIEW = 0
        val SUCCESS_FOCUS_VIEW = 1
        val FAILED_FOCUS_VIEW = 2
    }
}
